### DISCUSSION TITLE
Failed execution plugin

### DISCUSSION THREAD

**Discussion by TreF555** [elsa-workflows/elsa-core#1562] [Q&A]
Error while repeating example from https://elsa-workflows.github.io/elsa-core/docs/extensibility/extensibility-designer-plugins#dynamic-outcomes
![image](https://user-images.githubusercontent.com/2039365/134405643-22c7874d-3b3e-4487-bb69-a3b2ec246590.png)

**Activity.cs**

`[Action(Category = "Test",
        Outcomes = new[] { OutcomeNames.Done })]
    public class RandomizeOutcome : Activity
    {
        [ActivityInput(
             Hint = "Enter one or more outcome names.",
             UIHint = ActivityInputUIHints.MultiText,
             DefaultSyntax = SyntaxNames.Json,
             SupportedSyntaxes = new[] { SyntaxNames.Json },
             IsDesignerCritical = true
         )]
        public ISet<string> PossibleOutcomes { get; set; } = new HashSet<string>();
    }`

**Plugin**
`export default function RandomizeOutcomePlugin(elsaStudio) {
    const { eventBus } = elsaStudio;
    eventBus.on('activity-design-displaying', this.onActivityDesignDisplaying);

    this.onActivityDesignDisplaying = context => {
        alert('test');
        return;
        const activityModel = context.activityModel;

        // Only handle RandomizeOutcome activities.
        if (activityModel.type !== 'RandomizeOutcome')
            return;

        const props = activityModel.properties || [];
        const syntax = SyntaxNames.Json;

        // Get the value stored in the `PossibleOutcomes` property.
        // Keep in mind that activity properties are stored in a dictionary of expressions, keyed by syntax.
        const possibleOutcomes = props.find(x => x.name == 'PossibleOutcomes') || { expressions: { 'Json': '[]' }, syntax: syntax };
        const expression = possibleOutcomes.expressions[syntax] || [];

        // Set the outcomes property of the `context` argument to the list of possible outcomes.
        // Due to the way expressions are serialized depending on the syntax used, we need to check for a couple of formats in order to get the array.
        context.outcomes = !!expression['$values'] ? expression['$values'] : Array.isArray(expression) ? expression : parseJson(expression) || [];
    }
}`

**initialize plugIn**

`<script type="module">
        import  RandomizeOutcomePlugin  from "/js/test.js";
        // Get a handle to the elsa-studio-root element.
        const elsaStudioRoot = document.querySelector('elsa-studio-root');

        // Configure Elsa during the 'initializing' event.
        elsaStudioRoot.addEventListener('initializing', e => {
            const elsa = e.detail;
            //elsa.pluginManager.registerPlugin(TestPlugin);
            elsa.pluginManager.registerPlugin(RandomizeOutcomePlugin);
        });
    </script>`

**Version**
`<PackageReference Include="Elsa" Version="2.3.0-rc1" />
    <PackageReference Include="Elsa.Abstractions" Version="2.3.0-rc1" />
    <PackageReference Include="Elsa.Activities.Console" Version="2.3.0-rc1" />
    <PackageReference Include="Elsa.Activities.Http" Version="2.3.0-rc1" />
    <PackageReference Include="Elsa.Activities.Temporal.Common" Version="2.3.0-rc1" />
    <PackageReference Include="Elsa.Activities.Temporal.Quartz" Version="2.3.0-rc1" />
    <PackageReference Include="Elsa.Core" Version="2.3.0-rc1" />
    <PackageReference Include="Elsa.Designer.Components.Web" Version="2.3.0-rc1" />
    <PackageReference Include="Elsa.Persistence.EntityFramework.Sqlite" Version="2.3.0-rc1" />
    <PackageReference Include="Elsa.Server.Api" Version="2.3.0-rc1" />`

**Comment by sfmskywalker** [comment]
Can you send me a project demonstrating this issue?

**Comment by TreF555** [comment]
posted the project

https://github.com/TreF555/ElsaDashboardTest/tree/master/ElsaDashboardTest

How to repeat - create "Randomize Outcome", enter the tags test1 and test2 and save

**Reply by sfmskywalker** [reply]
Thanks for the repro, that was very helpful.

The issue is that the plugin registers a event handler that is undefined at that moment (because it is defined later in the code).

To fix that, simply reorder the declaration of your handler and the handler registration. Like this:

```javascript
export default function RandomizeOutcomePlugin(elsaStudio) {
    const { eventBus } = elsaStudio;
    
    // Declare handler first.
    this.onActivityDesignDisplaying = context => {
        // Code goes here...
    }

    // Then register.
    eventBus.on('activity-design-displaying', this.onActivityDesignDisplaying);
}
```

**Comment by TreF555** [comment]
Thanks for reply.
Error in console now not, and list PossibleOutcomes tags not view
commit https://github.com/TreF555/ElsaDashboardTest/tree/master/ElsaDashboardTest
![image](https://user-images.githubusercontent.com/2039365/135331542-a84d21f3-cf04-4eb9-996a-6fca1f2090ed.png)
After save activity
![image](https://user-images.githubusercontent.com/2039365/135331674-3c5d2333-1849-47b1-ba22-3b4bd3597b06.png)
Only Done

**Comment by TreF555** [comment]
Error in console http://localhost:33000/workflow-instances
![image](https://user-images.githubusercontent.com/2039365/135333554-5b2a34ad-c06a-4f01-b7b2-8646caf43092.png)

**Reply by sfmskywalker** [reply]
It works fine for me when I try:

![image](https://user-images.githubusercontent.com/938393/135848935-96d0e308-7c41-41cc-88fc-d0189ef83e34.png)

However I did encounter a few other issues for which I am pushing a fix shortly.
When these fixes are built, I recommend that you update to the latest prerelease packages available from MyGet and try again. If you are still seeing the issue, please let me know.

Thanks
